go-controller/pkg/metrics/master.go

```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+// It will return ovsdb operations which a user can add to existing operations they wish to track.
+// Upon successful transaction of the operations to the ovsdb server, the user of this function must call a call-back
+// function to lock-in the request to measure and report. Failure to call the call-back function, will result in no OVN
+// measurement and no metrics are reported. AddOVN will result in a no-op if Start isn't called previously for the same
+// kind/namespace/name and resource version.
+// If multiple AddOVN is called between Start and End for the same kind/namespace/name, then the
+// OVN durations will be summed and added to the total. There is an assumption that processing of kind/namespace/name is
+// sequential
+func (cr *ConfigDurationRecorder) AddOVN(nbClient libovsdbclient.Client, kind, namespace, name string) (
+	[]ovsdb.Operation, func(), time.Time, error) {
+	if !cr.enabled {
+		return []ovsdb.Operation{}, func() {}, time.Time{}, nil
+	}
+	kindNamespaceName := fmt.Sprintf("%s/%s/%s", kind, namespace, name)
+	if !cr.allowedToMeasure(kindNamespaceName) {
+		return []ovsdb.Operation{}, func() {}, time.Time{}, nil
+	}
+	cr.measurementsMu.RLock()
+	m, ok := cr.measurements[kindNamespaceName]
+	cr.measurementsMu.RUnlock()
+	if !ok {
+		// no measurement found, therefore no-op
+		return []ovsdb.Operation{}, func() {}, time.Time{}, nil
+	}
+	if m.end {
+		// existing measurement in-progress and not processed yet, therefore no-op
+		return []ovsdb.Operation{}, func() {}, time.Time{}, nil
+	}
+	nbGlobal := &nbdb.NBGlobal{}
+	nbGlobal, err := libovsdbops.GetNBGlobal(nbClient, nbGlobal)
+	if err != nil {
+		return []ovsdb.Operation{}, func() {}, time.Time{}, fmt.Errorf("failed to find OVN Northbound NB_Global table"+
+			" entry: %v", err)
+	}
+	if nbGlobal.NbCfg < 0 {
+		return []ovsdb.Operation{}, func() {}, time.Time{}, fmt.Errorf("nb_cfg is negative, failed to add OVN measurement")
+	}
+	//stop recording if we are close to overflow
+	if nbGlobal.NbCfg > maxNbCfg {
+		return []ovsdb.Operation{}, func() {}, time.Time{}, fmt.Errorf("unable to measure OVN due to nb_cfg being close to overflow")
+	}
+	ops, err := nbClient.Where(nbGlobal).Mutate(nbGlobal, model.Mutation{
+		Field:   &nbGlobal.NbCfg,
+		Mutator: ovsdb.MutateOperationAdd,
+		Value:   1,
+	})
+	if err != nil {
+		return []ovsdb.Operation{}, func() {}, time.Time{}, fmt.Errorf("failed to create update operation: %v", err)
+	}
+	ovnStartTimestamp := time.Now()
+
+	return ops, func() {
+		// there can be a race condition here where we queue the wrong nbCfg value, but it is ok as long as it is
+		// less than or equal the hv_cfg value we see and this is the case because of atomic increments for nb_cfg
+		cr.measurementsMu.Lock()
+		m, ok = cr.measurements[kindNamespaceName]
+		if !ok {
+			klog.V(5).Infof("Config duration recorder: expected a measurement entry. Call Start before AddOVN"+
```

Comment:
should this be an error? When is this OK?

Reply:
Yep. It should be an error and it should never occur.